### ISSUE TITLE
Visibility check is over-restrictive when a public method exists with the same name

### ISSUE THREAD

**Issue by mattrberry** [crystal-lang/crystal#10986]
When checking types of calls, it seems that the compiler does the [visibility check](https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/semantic/call.cr#L370) first and unconditionally, which causes it to error even when a valid public method exists.

```crystal
class C
  def foo(a : Bool | Int32)
    puts "public Bool | Int32"
  end
 
  private def foo(a : Int32)
    puts "private Int32"
  end
end
 
C.new.foo(true ? false : 1)
```
```
Error: private method 'foo' called for C
```

A use-case I could imagine for this would be for the public method to accept a union of a few inputs, then pass those along to private methods to handle each one.

Files where this behavior is tested: visibility_modifiers_spec, generic_class_spec, enum_spec, and class_spec

**Comment by straight-shoota** [comment]
I think the reason is just that `foo(a : Int32)` is more specific than `foo(a : Bool | Int32)` and wins as the overload for `foo(Int32)`. It's true that visibility is only checked afterwards. Maybe that could be changed. But I also think it might introduce confusion. AFAIK visibility modifiers currently have no influence on method resolution, they just inhibit a call if the selected overload is not visible from the calling scope.
This may not sound intuitive, but practically, there's not much difference actually and I presume it's easier to do (and reason about).

I don't see a strong motivation to change that.

A typical solution for the use case you described is naming the internal methods like `foo_internal`. That's IMO a superior approach (even if visibility modifiers worked the way you wished) because it clearly expresses the intention of these methods and the relation between public and private methods. It also prevents accidentally hitting the internal methods directly from the private scope.

> Files where this behavior is tested: visibility_modifiers_spec, generic_class_spec, enum_spec, and class_spec

What exactly do you mean by that?

**Comment by beta-ziliani** [comment]
@straight-shoota note that the argument is of type `Bool | Int32`. If you remove the `private` declaration, it picks the correct first `foo`. This seems like a bug to me.

**Comment by straight-shoota** [comment]
It picks the first one because the *actual* value of the argument is `Bool`. Try `false ? true : 1` and the second one is picket:

```cr
class C
  def foo(a : Bool | Int32)
    1
  end
 
  def foo(a : Int32)
    2
  end
end

C.new.foo(false ? true : 1) # => 2
```

**Comment by Sija** [comment]
> It picks the first one because the *actual* value of the argument is `Bool`. Try `false ? true : 1` and the second one is picked:
> 
> 
> 
> ```cr
> 
> class C
> 
>   def foo(a : Bool | Int32)
> 
>     1
> 
>   end
> 
>  
> 
>   def foo(a : Int32)
> 
>     2
> 
>   end
> 
> end
> 
> 
> 
> C.new.foo(false ? true : 1) # => 2
> 
> ```

@straight-shoota But it picks the 2nd one regardless, that's the whole issue.

**Comment by asterite** [comment]
> A use-case I could imagine for this would be for the public method to accept a union of a few inputs, then pass those along to private methods to handle each one.

Use a different name for the private ones?

**Comment by straight-shoota** [comment]
I understand that. My comment was an answer to beta's claim that removing `private` would make it pick the first overload always. Which overload is picked depends on the runtime type of the argument. The visibility checker runs for all compile-time types.

**Comment by beta-ziliani** [comment]
Perhaps there are strong technical reasons for this behavior, but to the eyes of a user this looks like a bug. Why is it considering private methods in the first place? If impossible to fix, I would rather disallow private overloads of public methods.

**Comment by straight-shoota** [comment]
> If impossible to fix, I would rather disallow private overloads of public methods.

That sounds overly restrictive really. For example, it's a relatively common pattern to define private constructors (besides public ones). In that case you can't even go to a different method name because `initialize` is provisioned by the compiler.

Similarly, I don't see any issues with overloads of a method having different visibilities.
Overriding could be a problem, though. Especially in type hierarchies.

**Comment by lbguilherme** [comment]
As a benchmark with C++: When there are two valid member functions, a public and a private one and both are valid candidates but the signature of the private one is a better match, the compiler picks the private one and errors.

I think this shouldn't be hard to change, but I question if we should.

**Comment by mattrberry** [comment]
FWIW, this behavior works in Java. I came across it in a Java program at work yesterday and figured I'd check how Crystal behaves. That's not to say that Crystal needs to follow Java, but just something to consider

**Comment by asterite** [comment]
Java doesn't have multiple dispatch, so I'm not sure it's equivalent there.

**Comment by mattrberry** [comment]
Right, my example above isn't a direct comparison to Java. I can elaborate.

Consider the following Java classes:
```java
abstract class A {}
class B extends A {}
class C extends A {}

class Foo {
    public void foo(final A a) {
        System.out.println("handle common logic");
        if (a instanceof B) {
            foo((B) a);
        } else if (a instanceof C) {
            foo((C) a);
        }
    }

    private void foo(final B b) {
        System.out.println("handle b");
    }

    private void foo(final C c) {
        System.out.println("handle c");
    }
}
```
In this case, when calling `new Foo().foo(new B())`, the following is printed to stdout:
```
handle common logic
handle b
```
In Crystal, I would translate that directly to the following (could use `.is_a?` in an if-elsif condition, but that's functionally equivalent, right?):
```crystal
abstract class A; end 
class B < A; end 
class C < A; end 

class Foo 
  def foo(a : A) : Nil
    puts "handle common logic"
    case a
    in B then foo(a)
    in C then foo(a)
    end
  end

  private def foo(b : B) : Nil 
    puts "handle b"
  end

  private def foo(c : C) : Nil 
    puts "handle c"
  end
end

Foo.new.foo(B.new)
```
However, the Crystal compiler complains about this with the error
```
Error: private method 'foo' called for Foo
```
The exact same holds true with union types:
```crystal
class Foo 
  def foo(a : Bool | Int32) : Nil 
    puts "handle common logic"
    case a
    in Bool  then foo(a)
    in Int32 then foo(a)
    end 
  end 

  private def foo(b : Bool) : Nil 
    puts "handle bool"
  end 

  private def foo(i : Int32) : Nil 
    puts "handle int32"
  end 
end

Foo.new.foo(true)
```
```
Error: private method 'foo' called for Foo
```
To me, this is not intuitive. The call to Foo#new from outside of the class should have no knowledge of the private methods. In terms of visibility, that should be equivalent the following, which obviously does compile:
```crystal
class Foo 
  def foo(a : Bool | Int32) : Nil
  end
end

Foo.new.foo(true)
```
In my original example, my expectation would be that the call to `#foo(a : Bool | Int32) : Nil` would hit the public method, since that's all that is visible. The public method can then delegate that anywhere else that it pleases, even if it's to an internal method with the same name and a narrower type.

**Comment by asterite** [comment]
Right, but this line:

```crystal
Foo.new.foo(B.new)
```

in Crystal it will call the `B` overload or the `C` overload depending on the runtime value of the type. This never happens in Java: only the compile-time value is used to determine the method to call.

That's why I'm saying that it's not fair to compare this with Java. In Java, when you call a method, there's always one match. If it's private, you can't call it. In Crystal, when you call a method, you might end up calling more than one method at runtime. If any of those is private, then you can't do it.

I don't see any inconsistency here. In Ruby if you try to call a private method, it tells you that you can't do that because it's private.

**Comment by asterite** [comment]
Maybe we can change the way things work: if you call a method "from the outside", then private methods are not looked up. However, I think that would bring a lot of confusion: if you call that same method from the inside it won't work exactly the same because those other private overloads will be considered.

**Comment by beta-ziliani** [comment]
The confusion exists anyway, so the point is which confusion are we OK with. In the end, it might just be a matter of improving error reporting.

**Comment by mattrberry** [comment]
My expectation would be that's how it works. A call from inside the class _should_ consider more methods than a call from outside, because those private methods are in scope at that point. That's obviously just my opinion/expectation, though, not to say that that has to be Crystal's approach

**Comment by HertzDevil** [comment]
There is already a predecent for private names overriding public names, for types:

```crystal
private class A
  class B; end
end

class A
end

{% A::B %} # okay
```

```crystal
class A
end

private class A
  class B; end
end

{% A::B %} # okay
```

```crystal
class A
  class B; end
end

private class A
end

{% A::B %} # Error: undefined constant A::B
```

```crystal
private class A
end

class A
  class B; end
end

{% A::B %} # Error: undefined constant A::B
```

and similarly, for constants:

```crystal
private A = 1
A = 2
A # => 1
```

```crystal
A = 1
private A = 2
A # => 2
```
